NetKet/Machine/sum_output.hpp

```diff
@@ -0,0 +1,119 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_SUMOUTPUT_HH
+#define NETKET_SUMOUTPUT_HH
+
+#include <time.h>
+#include <Eigen/Dense>
+#include <algorithm>
+#include <complex>
+#include <fstream>
+#include <random>
+#include <vector>
+#include "Lookup/lookup.hpp"
+#include "Utils/all_utils.hpp"
+#include "abstract_layer.hpp"
+
+namespace netket {
+
+template <typename T>
+class SumOutput : public AbstractLayer<T> {
+  using VectorType = typename AbstractLayer<T>::VectorType;
+  using MatrixType = typename AbstractLayer<T>::MatrixType;
+
+  int in_size_;   // input size: should be multiple of no. of sites
+  int out_size_;  // output size: should be multiple of no. of sites
+
+  VectorType z_;  // Linear term, z = W' * in + b
+
+  VectorType din_;  // Derivative of the input of this layer.
+                    // Note that input of this layer is also the output of
+                    // previous layer
+
+ public:
+  using StateType = typename AbstractLayer<T>::StateType;
+  using LookupType = typename AbstractLayer<T>::LookupType;
+
+  /// Constructor
+  explicit SumOutput(const json &pars) {
+    in_size_ = FieldVal(pars, "Inputs");
+
+    out_size_ = 1;
+
+    Init();
+  }
+
+  void Init() {
+    din_.resize(in_size_);
+    din_.setConstant(1);
+    z_.resize(out_size_);
+
+    std::string buffer = "";
+    InfoMessage(buffer) << "Sum Output Layer: " << in_size_ << " --> "
+                        << out_size_ << std::endl;
+  }
+
+  void InitRandomPars(int /*seed*/, double /*sigma*/) override {}
+
+  int Npar() const override { return 0; }
+
+  int Ninput() const override { return in_size_; }
+
+  int Noutput() const override { return out_size_; }
+
+  void GetParameters(VectorType & /*pars*/, int /*start_idx*/) override {}
+
+  void SetParameters(const VectorType & /*pars*/, int /*start_idx*/) override {}
+
+  void InitLookup(const Eigen::VectorXd & /*v*/, LookupType & /*lt*/) override {
+  }
+
+  void UpdateLookup(const Eigen::VectorXd & /*v*/,
+                    const std::vector<int> & /*tochange*/,
+                    const std::vector<double> & /*newconf*/,
+                    LookupType & /*lt*/) override {}
+
+  void Forward(const VectorType &prev_layer_data) override {
+    z_(0) = prev_layer_data.sum();
+  }
+
+  // Using lookup
+  void Forward(const VectorType &prev_layer_data,
+               const LookupType & /*lt*/) override {
+    z_(0) = prev_layer_data.sum();
+  }
+
+  VectorType Output() const override { return z_; }
+
+  void Backprop(const VectorType & /*prev_layer_data*/,
```

Comment:
Backprop is not implemented in this case? I know it does not make sense to backprob the final layer, but what if somebody wants this layer in the middle of a network?

Reply:
It would still work. There is no parameters in this layer so backprop doesnt need to do anything. It should still work if you put it in the middle.

Reply:
Ok sorry. It didnt work before but now it should be fine.

Reply:
Could you add test cases to check the back (and forward) propagation for all the layer types. If we have tests that catch errors such as the one here we could be a lot more confident that the implementation is correct.